deis/builder
# fix(pkg/sshd): wrap git receive functionality in a locking utility fu…

# Summary of Changes

This patch wraps the logic that runs the git receive hook in a utility function which calls `defer pushLock.Unlock()` so that, even if a `git push` fails, the lock gets unlocked.

# Associated End to End Tests

Created https://github.com/deis/workflow-e2e/issues/207 to track the work required to test this new behavior

# Issue(s) that this PR Closes

Closes #309
Note that this patch doesn't attempt to fix https://github.com/deis/builder/issues/220

# Testing Instructions

1. Create a Deis Cluster, specifying a dockerbuilder and slugbuilder image that doesn't exist (see [here](https://github.com/deis/charts/blob/master/workflow-dev/tpl/deis-builder-rc.yaml#L35-L38))
2. Register an app
3. Add a key to your app
4. `git push deis master` from a repository
5. Cancel the `git push` while it's running by pressing ctrl+c

After step 5, you should be able to do another `git push` without getting an error, but you may have to wait a few seconds.

# Pull Request Hygiene TODOs

Please make sure the below checklist is complete.
